pkg/watchdog/watchdog.go

```diff
@@ -0,0 +1,236 @@
+//     while running {
+//         wd.KeepAlive()
+//     }
+//     wd.MagicClose()
+//
+// Open() arms the watchdog. MagicClose() disarms the watchdog.
+//
+// Alternatively, use Close() which has behavior dependent on
+// CONFIG_WATCHDOG_NOWAYOUT.
+//
+// Note not every watchdog driver supports every function!
+//
+// For more, see:
+// https://www.kernel.org/doc/Documentation/watchdog/watchdog-api.txt
+package watchdog
+
+import (
+	"log"
+	"os"
+	"unsafe"
+
+	"golang.org/x/sys/unix"
+)
+
+// Dev is the name of the first watchdog. If there are multiple watchdogs, they
+// are named /dev/watchdog0, /dev/watchdog1, ...
+const Dev = "/dev/watchdog"
+
+// Various ioctl numbers.
+const (
+	wdiocGetSupport    = 0x80285700
+	wdiocGetStatus     = 0x80045701
+	wdiocGetBootStatus = 0x80045702
+	wdiocGetTemp       = 0x80045703
+	wdiocSetOptions    = 0x80045704
+	wdiocKeepAlive     = 0x80045705
+	wdiocSetTimeout    = 0xc0045706
+	wdiocGetTimeout    = 0x80045707
+	wdiocSetPreTimeout = 0xc0045708
+	wdiocGetPreTimeout = 0x80045709
+	wdiocGetTimeLeft   = 0x8004570a
+)
+
+// WatchdogInfo is returned by the GetSupport ioctl.
+type WatchdogInfo struct {
+	// Options the card/driver supports
+	options uint32
+	// Firmware version of the card
+	firmwareVersion uint32
+	// Identity of the board
+	identity [32]uint8
+}
+
+// Flags returned by GetStatus and GetBootStatus. These are the same flags used
+// for GetSupport()'s options field.
+const (
+	// Unknown flag error
+	StatusUnknown = -1
+	// Reset due to CPU overheat
+	StatusOverheat = 0x0001
+	// Fan failed
+	StatusFanFault = 0x0002
+	// External relay 1
+	StatusExtern1 = 0x0004
+	// ExStatusl relay 2
+	StatusExtern2 = 0x0008
+	// Power bad/power fault
+	StatusPowerUnder = 0x0010
+	// Card previously reset the CPU
+	StatusCardReset = 0x0020
+	// Power over voltage
+	StatusPowerOver = 0x0040
+	// Set timeout (in seconds)
+	StatusSetTimeout = 0x0080
+	// Supports magic close char
+	StatusMagicClose = 0x0100
+	// Pretimeout (in seconds), get/set
+	StatusPreTimeout = 0x0200
+	// Watchdog triggers a management or other external alarm not a reboot
+	StatusAlarmOnly = 0x0400
+	// Keep alive ping reply
+	StatusKeepAlivePing = 0x8000
+)
+
+// Settings passed to SetOptions.
+const (
+	// Unknown status error
+	OptionUnknown = -1
+	// Turn off the watchdog timer
+	OptionDisableCard = 0x0001
+	// Turn on the watchdog timer
+	OptionEnableCard = 0x0002
+	// Kernel panic on temperature trip
+	OptionTempPanic = 0x0004
+)
+
+type Watchdog struct {
+	f *os.File
+}
+
+func Open(dev string) (*Watchdog, error) {
+	f, err := os.OpenFile(dev, os.O_RDWR, 0)
+	if err != nil {
+		return nil, err
+	}
+	return &Watchdog{f: f}, nil
+}
+
+// Close has behavior dependent on CONFIG_WATCHDOG_NOWAYOUT.
+func (w *Watchdog) Close() error {
+	return w.Close()
+}
+
+// MagicClose disarms the watchdog.
+func (w *Watchdog) MagicClose() error {
+	if _, err := w.f.Write([]byte("V")); err != nil {
+		return err
+	}
+	return w.Close()
```

Comment:
you said above somewhere to use MagicClose instead of Close if you don't have WATCHDOG_NOWAYOUT, but here you use Close anyway? So it doesn't fail if you don't have that config turned on, but it just is ineffective?

Reply:
This is in the `MagicClose` function. If I called MagicClose here, it'd be infinite recursion. The magical part of magic close is it writes the special magic byte 'V' which gives you a way out of closing the watchdog fd without rebooting your system.